specs/pluggable_alert_delivery.adoc

```diff
@@ -0,0 +1,372 @@
+= Pluggable Alerts Delivery End-Points
+
+== Introduction
+
+The alerting module in tendrl needs to alert/notify
+
+* State changes of tendrl managed entities
+* Threshold breaches of utilizations of tendrl managed entities.
+
+via mediums such as:
+
+* SNMP-traps
+* email
+
+The different sources of alerts are:
+
+* Performance monitoring threshold breach alerts generated by collectd.
+* Alerts related to state changes of tendrl managed sds specific entities
+  generated by sds-integration.
+  ex: Tendrl/gluster-integration, Tendrl/ceph-integration
+
+== Problem description
+
+* The different tendrl supported alerting means need to be pluggable in the
+  sense that adding any new alert notifying means should be an easy task and
+  should not involve much changes in the core alerting framework.
+* The alerting mechanism should have a way to set importance of alerts so that
+  the alerts desingated as important will be notified to users even if the
+  alerting module crashed and hence the alert could not be notified but will
+  be alerted once the alerting module is started back.
+
+== Use Cases
+
+* Dynamic addition of new means of notifying an alert.
+
+== Proposed change
+
+* The alerts raised are persisted in '/alerts' directory of etcd before sending
+  out notifications. The flow is as below:
+  ** One socket is connected to for writes from collectd, read-only from the
+     node agent and read-write from the bridge.
+  ** The generator(collectd or bridge) of the alert needs to put the alert on
+     this socket.
+  ** Any alert on that socket is always taken to etcd by the node agent
+  ** Bridge can read the alerts and act on only the ones it can act on, ignore
+     the rest it can also generate it's own alerts and put them on the socket
+     for node agent to transport to etcd
+* Alert meta-data:
+  ** Each alert will carry additional attribute that carries its significance
+     which will be added by respective handler. The supported significance
+     levels can be 'HIGH', 'MEDIUM' and 'LOW'.
+    *** The 'LOW' significance level alerts will not be notified. But will
+        feature in the list of alerts in etcd.
+    *** The 'MEDIUM' significance level alerts will be attempted to be notified
+    *** The 'HIGH' significance level alerts will be re-attempted to be
+        notified(on restart) if a previous attempt failed for some reason.
+        (crash of alerting service).The alerts marked as significance HIGH are
+        deemed important and hence their delivery is guranteed even in case of
+        sudden crash of alerting module. This is achieved by updating a field
+        ('notified': True) of the alert in etcd directory to indicate that the
```

Comment:
Given the pluggable nature of our transports and the possibility that each alert may need to be delivered via more than one transport.

The alerting module is updating the alerts in any case. So, I think we can expand that a bit and have a field called `delivery` or some such added by default to each alert. The alerting module would add individual endpoints there, for example:

`
delivery: {
  smtp: {
    endpoint: 'foo@bar.com',
    pickedup: True,
    delivered: False
  },
  snmp: {
     endpoint: 'foobar',
     pickedup: True,
     delivered: False
  }
}
`

This brings up another point. How do we determine if there are unsent alerts? That's where the `notified` field you've used comes in. If you notice the `pickedup` fields above, they're set to True. The reason is that I'm assuming that the alerting module picks up each alert as soon as it comes across it. At which point it adds the entire `delivery` field. At this point I'm also assuming that it immediately invokes all the applicable delivery modules to deliver this alert. `pickedup` indicate that this was done. `delivered` is updated once the delivery module confirms an alert to have been successfully delivered. Each time the alerting module goes to update the value of the `delivered` key for a specific delivery module, it can update the global value if all the nested `delivered` as True.

There are additional concerns I can think of at this point. Are the delivery modules just point in time invocations, so that they have no state of their own? In which case, how do you determine the last attempted delivery time? This goes towards the question of figuring out which alerts need to be redelivered. Of course, I think this wouldn't be a problem if every delivery is a blocking, atomic call. It either succeeds or fails. In any case, you might want to put some interval before attempting redelivery and track the number of delivery attempts made as well.